src/win32/w32_stack.c

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#if defined(GIT_MSVC_CRTDBG)
+#include "Windows.h"
+#include "Dbghelp.h"
+#include "win32/posix.h"
+#include "w32_stack.h"
+#include "hash.h"
+
+/**
+ * This is supposedly defined in WinBase.h (from Windows.h) but there were linker issues.
+ */
+USHORT WINAPI RtlCaptureStackBackTrace(ULONG, ULONG, PVOID*, PULONG);
+
+static bool   g_win32_stack_initialized = false;
+static HANDLE g_win32_stack_process = INVALID_HANDLE_VALUE;
+
+void git_win32__stack_init(void)
+{
+	if (!g_win32_stack_initialized) {
+		g_win32_stack_process = GetCurrentProcess();
+		SymSetOptions(SYMOPT_LOAD_LINES);
+		SymInitialize(g_win32_stack_process, NULL, TRUE);
+		g_win32_stack_initialized = true;
+	}
+}
+
+void git_win32__stack_cleanup(void)
+{
+	if (g_win32_stack_initialized) {
+		SymCleanup(g_win32_stack_process);
+		g_win32_stack_process = INVALID_HANDLE_VALUE;
+		g_win32_stack_initialized = false;
+	}
+}
+
+int git_win32__stack_capture(git_win32__stack__raw_data *pdata, int skip)
+{
+	if (!g_win32_stack_initialized) {
+		giterr_set(GITERR_INVALID, "git_win32_stack not initialized.");
+		return GIT_ERROR;
+	}
+
+	memset(pdata, 0, sizeof(*pdata));
+	pdata->nr_frames = RtlCaptureStackBackTrace(
+		skip+1, GIT_WIN32__STACK__MAX_FRAMES, pdata->frames, NULL);
+
+	return 0;
+}
+
+int git_win32__stack_compare(
+	git_win32__stack__raw_data *d1,
+	git_win32__stack__raw_data *d2)
+{
+	return memcmp(d1->frames, d2->frames, sizeof(d1->frames));
+}
+
+int git_win32__stack_format(
+	char *pbuf, int buf_len,
+	const git_win32__stack__raw_data *pdata,
+	const char *prefix, const char *suffix)
+{
+#define MY_MAX_FILENAME 255
+
+	/* SYMBOL_INFO has char FileName[1] at the end.  The docs say to
+	 * to malloc it with extra space for your desired max filename.
+	 * We use a union to do the same thing without mallocing.
+	 */
+	union {
+		SYMBOL_INFO symbol;
+		char buf[sizeof(SYMBOL_INFO) + MY_MAX_FILENAME + 1];
+	} s;
```

Comment:
I think it'd be a bit more idiomatic libgit2 to declare our own struct with `SYMBOL_INFO` as the first field and the buffer just after that, though I suppose we don't really do quite this anywhere else.